### PR TITLE
Pin analyzers that ship in the SDK

### LINKED ISSUE #36552: Building a 5.0 app with 6.0 SDK in VS 2019 causes analyzers to fail to load
MVC and StartupAnalyzer are shipped as part of the SDK and target every version of .NET 3.1 onwards. These analyzers have been updated to reference Microsoft.CodeAnalysis versions 4.0.0.0 which isn't available on VS 2019. Consequently opening a 5.0 project with a 6.0 SDK results in a bunch of warnings:

```
1>CSC : warning CS8032: An instance of analyzer Microsoft.AspNetCore.Analyzers.StartupAnalyzer cannot be created from C:\Program Files\dotnet\sdk\6.0.100-rc.2.21464.10\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs\Microsoft.AspNetCore.Analyzers.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Microsoft.AspNetCore.Mvc.Analyzers.AttributesShouldNotBeAppliedToPageModelAnalyzer cannot be created from C:\Program Files\dotnet\sdk\6.0.100-rc.2.21464.10\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs\Microsoft.AspNetCore.Mvc.Analyzers.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Microsoft.AspNetCore.Mvc.Analyzers.AvoidHtmlPartialAnalyzer cannot be created from C:\Program Files\dotnet\sdk\6.0.100-rc.2.21464.10\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs\Microsoft.AspNetCore.Mvc.Analyzers.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Microsoft.AspNetCore.Mvc.Analyzers.TagHelpersInCodeBlocksAnalyzer cannot be created from C:\Program Files\dotnet\sdk\6.0.100-rc.2.21464.10\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs\Microsoft.AspNetCore.Mvc.Analyzers.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Microsoft.AspNetCore.Mvc.Analyzers.TopLevelParameterNameAnalyzer cannot be created from C:\Program Files\dotnet\sdk\6.0.100-rc.2.21464.10\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs\Microsoft.AspNetCore.Mvc.Analyzers.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Microsoft.AspNetCore.Components.Analyzers.ComponentParameterAnalyzer cannot be created from C:\Program Files\dotnet\sdk\6.0.100-rc.2.21464.10\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs\Microsoft.AspNetCore.Components.Analyzers.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Microsoft.AspNetCore.Components.Analyzers.ComponentParameterUsageAnalyzer cannot be created from C:\Program Files\dotnet\sdk\6.0.100-rc.2.21464.10\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs\Microsoft.AspNetCore.Components.Analyzers.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Microsoft.Extensions.Internal.ComponentInternalUsageDiagnosticAnalyzer cannot be created from C:\Program Files\dotnet\sdk\6.0.100-rc.2.21464.10\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs\Microsoft.AspNetCore.Components.Analyzers.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Microsoft.AspNetCore.Analyzers.StartupAnalyzer cannot be created from C:\Program Files\dotnet\sdk\6.0.100-rc.2.21464.10\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs\Microsoft.AspNetCore.Analyzers.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Microsoft.AspNetCore.Mvc.Analyzers.AttributesShouldNotBeAppliedToPageModelAnalyzer cannot be created from C:\Program Files\dotnet\sdk\6.0.100-rc.2.21464.10\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs\Microsoft.AspNetCore.Mvc.Analyzers.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Microsoft.AspNetCore.Mvc.Analyzers.AvoidHtmlPartialAnalyzer cannot be created from C:\Program Files\dotnet\sdk\6.0.100-rc.2.21464.10\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs\Microsoft.AspNetCore.Mvc.Analyzers.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Microsoft.AspNetCore.Mvc.Analyzers.TagHelpersInCodeBlocksAnalyzer cannot be created from C:\Program Files\dotnet\sdk\6.0.100-rc.2.21464.10\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs\Microsoft.AspNetCore.Mvc.Analyzers.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Microsoft.AspNetCore.Mvc.Analyzers.TopLevelParameterNameAnalyzer cannot be created from C:\Program Files\dotnet\sdk\6.0.100-rc.2.21464.10\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs\Microsoft.AspNetCore.Mvc.Analyzers.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Microsoft.AspNetCore.Components.Analyzers.ComponentParameterAnalyzer cannot be created from C:\Program Files\dotnet\sdk\6.0.100-rc.2.21464.10\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs\Microsoft.AspNetCore.Components.Analyzers.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Microsoft.AspNetCore.Components.Analyzers.ComponentParameterUsageAnalyzer cannot be created from C:\Program Files\dotnet\sdk\6.0.100-rc.2.21464.10\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs\Microsoft.AspNetCore.Components.Analyzers.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Microsoft.Extensions.Internal.ComponentInternalUsageDiagnosticAnalyzer cannot be created from C:\Program Files\dotnet\sdk\6.0.100-rc.2.21464.10\Sdks\Microsoft.NET.Sdk.Web\analyzers\cs\Microsoft.AspNetCore.Components.Analyzers.dll : Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified..
```

We should pin these to the the most recent version supported by VS2019.

### PR DESCRIPTION
**Pull request by pranavkm**
ASP.NET Core produces a few analyzer assemblies that are shipped as part of the .NET SDK. These analyzers are added to web projects targeting 3.1 and newer.
In 6.0, we (unintentionally) bumped the version of Microsoft.CodeAnalysis that these projects referenced to a 4.0 version. This causes warnings when opening
a 3.1 or 5.0 app in VS 2019 as it does not support these versions.

Additionally update the version of Microsoft.CodeAnalysis.* packages used in tests, Razor, and framework analyzers that are only expected to run with VS 2020 to a more recent build. This is largely book-keeping, but allows us
to write a test for file scoped namespaces.

Fixes https://github.com/dotnet/aspnetcore/issues/36552

### CODE REVIEW

File: src/Analyzers/Analyzers/test/MinimalStartupTest.cs [hunk 1]
```diff
@@ -53,6 +53,9 @@ public async Task StartupAnalyzer_WorksWithNonImplicitMain()
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.AspNetCore.Authorization;
 using Microsoft.AspNetCore.Mvc;
+
+namespace MyApp;
```
Comment (pranavkm) [inline src/Analyzers/Analyzers/test/MinimalStartupTest.cs:57]: FYI @BrennanConroy

Comment (BrennanConroy) [reply src/Analyzers/Analyzers/test/MinimalStartupTest.cs:57]: ?

File: eng/Versions.props [hunk 2]
```diff
@@ -188,9 +188,16 @@
...
+    <!--
+      Versions of Microsoft.CodeAnalysis packages referenced by analyzers shipped in the SDK.
+      This need to be pinned since they're used in 3.1 apps and need to be loadable in VS 2019.
+    -->
+    <Analyzer_MicrosoftCodeAnalysisCSharpVersion>3.3.1</Analyzer_MicrosoftCodeAnalysisCSharpVersion>
+    <Analyzer_MicrosoftCodeAnalysisCSharpWorkspacesVersion>3.3.1</Analyzer_MicrosoftCodeAnalysisCSharpWorkspacesVersion>
+
+    <MicrosoftCodeAnalysisCommonVersion>4.0.0-3.final</MicrosoftCodeAnalysisCommonVersion>
+    <MicrosoftCodeAnalysisCSharpVersion>4.0.0-3.final</MicrosoftCodeAnalysisCSharpVersion>
+    <MicrosoftCodeAnalysisCSharpWorkspacesVersion>4.0.0-3.final</MicrosoftCodeAnalysisCSharpWorkspacesVersion>
```
Comment (dougbu) [inline eng/Versions.props:200]: 1. /fyi these versions and the previous ones are highly unlikely to work once we stabilize for 6.0.0. I discussed this offline w/ @jaredpar and he's got a plan in place 😃
2. There's already `4.0.0-4.final` packages available on NuGet.org

**Review by dougbu — APPROVED** [review]
I'm missing something: How would we end up inserting a 7.0 analyzer into VS 2019❔ Even `net6.0` is only supported in VS 2022.

I suspect the answer is that we need to back-port this to a servicing branch. But, which one(s)❔

File: src/Analyzers/Analyzers/test/MinimalStartupTest.cs (hunk 1 above)
Comment (pranavkm) [reply src/Analyzers/Analyzers/test/MinimalStartupTest.cs:57]: It's using a file scoped namespace. https://github.com/dotnet/aspnetcore/issues/36552 called out that the logging source generator had trouble with it, and I was doing a little bit of sanity testing.

File: eng/Versions.props (hunk 2 above)
Comment (pranavkm) [inline eng/Versions.props:200]:
```suggestion
    <MicrosoftCodeAnalysisCommonVersion>4.0.0-4.final</MicrosoftCodeAnalysisCommonVersion>
    <MicrosoftCodeAnalysisCSharpVersion>4.0.0-4.final</MicrosoftCodeAnalysisCSharpVersion>
    <MicrosoftCodeAnalysisCSharpWorkspacesVersion>4.0.0-4.final</MicrosoftCodeAnalysisCSharpWorkspacesVersion>
```

**Comment by pranavkm** [comment]
@doubu these analyzers are added to the WebSDK and apply to 3.1 and later: https://github.com/dotnet/sdk/blob/main/src/WebSdk/Web/Sdk/Sdk.props#L38-L59. So by installing a 6.0 SDK, they'll start getting used in a user's app to build their 5.0 projects once 6.0 becomes RTM or if they have preview SDK feature checked. My plan is to port this to 6.0 next and possibly to 6.0-rc2.

**Pull request merged**